drivers/sensor/st/lsm9ds1_mag/lsm9ds1_mag.c

```diff
@@ -0,0 +1,407 @@
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm9ds1_ctrl_reg2_m_t ctrl_reg2;
+	int ret;
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG2_M, (uint8_t *)&ctrl_reg2, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctrl_reg2.reboot = 1;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG2_M, (uint8_t *)&ctrl_reg2, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_busy_wait(USEC_PER_MSEC * 50U);
+
+	return 0;
+}
+
+static int lsm9ds1_mag_range_to_fs_val(int32_t range)
+{
+	switch (range) {
+	case 4:
+		return LSM9DS1_4Ga;
+	case 8:
+		return LSM9DS1_8Ga;
+	case 12:
+		return LSM9DS1_12Ga;
+	case 16:
+		return LSM9DS1_16Ga;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lsm9ds1_mag_range_set(const struct device *dev, int32_t range)
+{
+	const struct lsm9ds1_mag_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_mag_data *data = dev->data;
+	int fs, ret;
+
+	fs = lsm9ds1_mag_range_to_fs_val(range);
+
+	ret = lsm9ds1_mag_full_scale_set(ctx, fs);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->mag_gain = lsm9ds1_mag_fs_sens[fs];
+
+	return 0;
+}
+
+static int lsm9ds1_mag_freq_to_odr_val(uint16_t freq)
+{
+	for (int i = 0; i < ARRAY_SIZE(lsm9ds1_mag_odr_map); i++) {
+		if (freq <= lsm9ds1_mag_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_mag_freq_to_fast_odr_val(uint16_t freq)
+{
+	for (int i = ARRAY_SIZE(lsm9ds1_mag_fast_odr_map) - 1; i >= 0; i--) {
+		if (freq <= lsm9ds1_mag_fast_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_mag_odr_set_normal(const struct device *dev, uint16_t freq)
+{
+	const struct lsm9ds1_mag_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_mag_data *data = dev->data;
+	lsm9ds1_ctrl_reg1_m_t ctrl_reg1_m;
+	lsm9ds1_ctrl_reg4_m_t ctrl_reg4_m;
+	int odr, ret;
+
+	odr = lsm9ds1_mag_freq_to_odr_val(freq);
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG1_M, (uint8_t *)&ctrl_reg1_m, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (ctrl_reg1_m.fast_odr) { /* restore the operating mode */
+		ctrl_reg1_m.om = data->old_om;
+
+		ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG4_M, (uint8_t *)&ctrl_reg4_m, 1);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ctrl_reg4_m.omz = data->old_om;
+
+		ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG4_M, (uint8_t *)&ctrl_reg4_m, 1);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ctrl_reg1_m._do = odr;
+	ctrl_reg1_m.fast_odr = 0;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG1_M, (uint8_t *)&ctrl_reg1_m, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int lsm9ds1_mag_fast_odr_set(const struct device *dev, uint16_t freq)
+{
+	const struct lsm9ds1_mag_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_mag_data *data = dev->data;
+	lsm9ds1_ctrl_reg1_m_t ctrl_reg1_m;
+	lsm9ds1_ctrl_reg4_m_t ctrl_reg4_m;
+	int odr, ret;
+
+	odr = lsm9ds1_mag_freq_to_fast_odr_val(freq);
```

Comment:
odr can be `-EINVAL` - shouldn't this be checked?

Reply:
Done